src/kokkos/plugin-PixelVertexFinding/kokkos/gpuClusterTracksByDensity.h

```diff
@@ -0,0 +1,257 @@
+#ifndef RecoPixelVertexing_PixelVertexFinding_src_gpuClusterTracksByDensity_h
+#define RecoPixelVertexing_PixelVertexFinding_src_gpuClusterTracksByDensity_h
+
+#ifdef TODO
+#include "CUDACore/HistoContainer.h"
+#include "CUDACore/cuda_assert.h"
+#endif  // TODO
+
+#include "gpuVertexFinder.h"
+
+namespace KOKKOS_NAMESPACE {
+  namespace gpuVertexFinder {
+    // this algo does not really scale as it works in a single block...
+    // enough for <10K tracks we have
+    //
+    // based on Rodrighez&Laio algo
+    //
+    KOKKOS_INLINE_FUNCTION void clusterTracksByDensity(
+        Kokkos::View<ZVertices*, KokkosExecSpace> vdata,
+        Kokkos::View<WorkSpace*, KokkosExecSpace> vws,
+        int minT,       // min number of neighbours to be "seed"
+        float eps,      // max absolute distance to cluster
+        float errmax,   // max error to be "seed"
+        float chi2max,  // max normalized distance to cluster
+        const Kokkos::TeamPolicy<KokkosExecSpace>::member_type& team_member) {
+      constexpr bool verbose = false;  // in principle the compiler should optmize out if false
+
+      auto id = team_member.league_rank() * team_member.team_size() + team_member.team_rank();
+
+      if (verbose && 0 == id)
+        printf("params %d %f %f %f\n", minT, eps, errmax, chi2max);
+
+      auto er2mx = errmax * errmax;
+
+      auto& __restrict__ data = *vdata.data();
+      auto& __restrict__ ws = *vws.data();
+
+      auto nt = ws.ntrks;
+      float const* __restrict__ zt = ws.zt;
+      float const* __restrict__ ezt2 = ws.ezt2;
+
+      uint32_t& nvFinal = data.nvFinal;
+      uint32_t& nvIntermediate = ws.nvIntermediate;
+
+      uint8_t* __restrict__ izt = ws.izt;
+      int32_t* __restrict__ nn = data.ndof;
+      int32_t* __restrict__ iv = ws.iv;
+
+#ifdef TODO
+      assert(vdata.data());
+      assert(zt);
+
+      using Hist = HistoContainer<uint8_t, 256, 16000, 8, uint16_t>;
+
+      // Get shared team allocations on the scratch pad
+      Hist* hist = (Hist*)team_member.team_shmem().get_shmem(sizeof(Hist));
+      Hist::Counter* hws = (Hist::Counter*)team_member.team_shmem().get_shmem(sizeof(Hist::Counter) * 32);
```

Comment:
Please use `static_cast` (or `reinterpret_cast` if `static_cast` is not enough).

It's a bit pity that CUDA's
```cpp
    __shared__ Hist hist;
    __shared__ typename Hist::Counter hws[32];
```
is more type safe.